hypervisor/src/mshv/mod.rs

```diff
@@ -1573,17 +1506,56 @@ impl MshvVcpu {
         ghcb_gpa: u64,
     ) -> std::result::Result<cpu::VmExit, cpu::HypervisorCpuError> {
         // Clear the SW_EXIT_INFO1 register to indicate no error
-        let mut swei1_rw_gpa_arg = mshv_bindings::mshv_read_write_gpa {
-            base_gpa: ghcb_gpa + GHCB_SW_EXITINFO1_OFFSET,
-            byte_count: std::mem::size_of::<u64>() as u32,
-            ..Default::default()
-        };
-        self.fd
-            .gpa_write(&mut swei1_rw_gpa_arg)
-            .map_err(|e| cpu::HypervisorCpuError::GpaWrite(e.into()))?;
+        self.gpa_write(ghcb_gpa + GHCB_SW_EXITINFO1_OFFSET, &[0; 4])?;
 
         Ok(cpu::VmExit::Ignore)
     }
+
+    #[cfg(feature = "sev_snp")]
+    fn gpa_read(
+        &self,
+        gpa: u64,
+        data: &mut [u8],
+    ) -> std::result::Result<(), cpu::HypervisorCpuError> {
+        const CHUNK_SIZE: usize = 16;
+
+        for (gpa, chunk) in (gpa..).step_by(CHUNK_SIZE).zip(data.chunks_mut(CHUNK_SIZE)) {
+            let mut rw_gpa_arg = mshv_bindings::mshv_read_write_gpa {
+                base_gpa: gpa,
+                byte_count: chunk.len() as u32,
+                ..Default::default()
+            };
+            self.fd
+                .gpa_read(&mut rw_gpa_arg)
+                .map_err(|e| cpu::HypervisorCpuError::GpaRead(e.into()))?;
+
+            chunk.copy_from_slice(&rw_gpa_arg.data[..chunk.len()]);
+        }
+
+        Ok(())
+    }
+
+    #[cfg(feature = "sev_snp")]
+    fn gpa_write(&self, gpa: u64, data: &[u8]) -> std::result::Result<(), cpu::HypervisorCpuError> {
```

Comment:
Also formatting here looks bit inconsistent with gpa_read side of the code. All the parameters are in single line vs multiline in previous scenario.

Reply:
That wasn't a conscious choice, that's just how rustfmt decided to wrap. I changed the return type to `cpu::Result<()>` and now both signatures are short enough to fit on a single line :)